### ISSUE TITLE
Transmission Matrix Not Converting to Phylo Object

### ISSUE THREAD

**Issue by thednainus** [statnet/EpiModel#482]
Hello, 
I'm trying to write some code for HIV dynamics and I tried to save the transmat. However, when I try to convert the transmat to a phylogenetic tree using as.phylo.transmat, I seem to be going to a forever loop. 
I tried the example for covid here (http://statnet.org/nme/d5-s4-COVIDDemog.html) just to make sure it was not anything that I did not code correctly. However, when I get the transmat, I also end up into a loop that never ends for some values of v_sub (in the function as.phylo.transmat). Do you know what might be the cause?

Thanks,
Fabricia.

**Comment by smjenness** [comment]
Also tagging @skyebend here, as he was the one to originally develop this conversion code, several years ago. Skye if you have any quick impressions, they would be appreciated.

**Comment by thednainus** [comment]
Thanks for transferring this to EpiModel; I thought I posted that to the right R package (not sleeping much as my toddler is keeping me up at night)... I am new to this type of network modelling. For me to try to find a solution, I need to understand why a loop is being generated... Is this because of a cyclic graph?

**Comment by thednainus** [comment]
I've been trying to fix this bug. It might be something else that is causing it and _maybe_ not something wrong with the as.phylo.transmat function .

I did a few tests, and I realized that sometimes a same susceptible gets infected more than once in different time points. If I understood the code correctly that is not possible, right? Because we are looking for pairs of _infected_ and _not infected_ for a transmission to happen.

An example of a transmission matrix in which show a same susceptible getting infected more than once in different time points. Do you think that this could be the cause? If we don't expect this to happen, where exactly should I be looking to fix it?
I did check whether 1 was being assigned to newly infected individuals (and that seems to be the case, or maybe I am missing something).

```
     **at    sus     inf     transProb    actRate    finalProb**
107     575    652       0.1              3              0.271
97      575     913       0.1             3               0.271 
```

**Comment by thednainus** [comment]
I think I found the bug. Unfortunately, I just don't think I will have the time to update EpiModel and do a pull request.
I think the actual problem is in the function `[discord_edgelist](https://github.com/statnet/EpiModel/blob/dc3640500dc0b6a1fb9f7dc8a68d73490abcfe23/R/net.mod.infection.R#L303)`.
What is happening is that everything related to susceptible and infected should be using index as in for example:

Instead of `[stat <- matrix(status[el], ncol = 2)](https://github.com/statnet/EpiModel/blob/dc3640500dc0b6a1fb9f7dc8a68d73490abcfe23/R/net.mod.infection.R#L318)`

It should be:
```
uid <- dat$attr$uid
indexes <- match(c(el), uid)
stat <- matrix(status[indexes], ncol = 2)
```

The same for lines 325 and 326 that should be something like:
```
        sus <- pairs[, 1]
        inf <- pairs[, 2]
        del <- data.frame(at, sus, inf)
        sus_index <- match(sus, uid)
        inf_index <- match(inf, uid)
        keep <- rowSums(matrix(c(active[sus_index], active[inf_index]), ncol = 2)) == 2
```

And then back to whichever functions are calling` discord_edgelist` function, indexes should also be used like in here:
`[del$infDur <- at - infTime[del$inf]](https://github.com/statnet/EpiModel/blob/dc3640500dc0b6a1fb9f7dc8a68d73490abcfe23/R/net.mod.infection.R#L69)`

It should be something like, so as the transmission matrix is correct:
```
inf_indexes <- match(del$inf, uid)
del$infDur <- at - inf.time[inf_indexes]
```

**Comment by thednainus** [comment]
You also need to update other lines as well that call directly `del$inf` or `del$sus` like in here, for example:

https://github.com/statnet/EpiModel/blob/dc3640500dc0b6a1fb9f7dc8a68d73490abcfe23/R/net.mod.infection.R#L97

It should be something similar to:
 ```
      idsNewInf <- unique(del$sus)
      index_idsNewInf <- match(idsNewInf, uid)
      idsTrans <- unique(del$inf)
      index_idsTrans <- match(idsTrans, uid)

     status[index_idsNewInf] <- "i"
```

and so on...

**Comment by smjenness** [comment]
Yes, that makes sense. We had not designed the transmission matrix output to work with models with mortality, so the unique id (UID) was not necessary. But we should add it, and I appreciate your suggestions above for how to do so. We'll try to get to this in the next couple weeks. Thanks!

**Comment by smjenness** [comment]
@thednainus : this problem has now been fixed with #488. We implemented a general framework for unique (persistent) IDs for open-population models that gets used by `set_transmat` in the infection modules. No updates are needed for the "built-in" models, but for extension models like the COVID model above that you ran, it is necessary to update the births/arrivals module to use our new `append_core_attr` function that handles all of the unique ID work. We have updated the NME workshop example page, and here is a direct link to the new syntax: http://statnet.org/nme/d5-s4-COVIDDemog.html#Births. I also added an example of extraction of the transmission matrix and plot to a phylogram at the end. Please let me know if you have any further questions!
